help_function/src/onedpl_test_group_load.cpp

```diff
@@ -0,0 +1,205 @@
+  {
+     for (int i = 0; i < 512; ++i) {
+      if (ptr[i] != i) {
+        std::cout << func_name << "_blocked" <<" failed\n";
+        std::ostream_iterator<int> Iter(std::cout, ", ");
+        std::copy(ptr, ptr + 512, Iter);
+        std::cout << std::endl;
+        return false;
+        }
+     }
+     std::cout << func_name << "_blocked" <<" pass\n";
+    
+  }
+    
+  else{
+    int expected[512];
+    int num_threads = 128;
+    int items_per_thread = 4;
+    for (int i = 0;i < num_threads; ++i){
+        for(int j=0;j < items_per_thread; ++j){
+          expected[i * items_per_thread +j] = j * num_threads +i;  
+        }
+      }
+    for (int i = 0; i < 512; ++i) {
+        if (ptr[i] != expected[i]) {
+          std::cout << func_name << "_striped" <<" failed\n";
+          std::ostream_iterator<int> Iter(std::cout, ", ");
+          std::copy(ptr, ptr + 512, Iter);
+          std::cout << std::endl;
+          return false;
+        }
+      }
+  
+    std::cout << func_name << "_striped" <<" pass\n";
+   
+  }
+  return true;
+}
+
+bool subgroup_helper_validation_function(const int* ptr,const uint32_t *sg_sz, const char* func_name){
+  int expected[512];
+  int num_threads = 128;
+  int items_per_thread = 4;
+  uint32_t sg_sz_val = *sg_sz;
+  for (int i = 0; i < num_threads; ++i) {
+      for (int j = 0; j < items_per_thread; ++j) {
+          expected[items_per_thread * i + j] = (i / sg_sz_val) * sg_sz_val * items_per_thread + sg_sz_val * j + i % sg_sz_val;
+      }
+   }
+  for (int i = 0; i < 512; ++i) {
+    if (ptr[i] != expected[i]) {
+      std::cout <<" failed\n";
+      std::ostream_iterator<int> Iter(std::cout, ", ");
+      std::copy(ptr, ptr + 512, Iter);
+      std::cout << std::endl;
+      return false;
+    }
+  }
+
+  std::cout << func_name <<" pass\n";
+  return true;
+}
+
+template<dpct::group::load_algorithm T>
+bool test_group_load() {
+  // Tests dpct::group::load_algorithm::BLOCK_LOAD_DIRECT & dpct::group::load_algorithm::BLOCK_LOAD_STRIPED 
+  // in its entirety as API functions
+  sycl::queue q;
+  oneapi::dpl::counting_iterator<int> count_it(0);
+  sycl::buffer<int, 1> buffer(count_it, count_it + 512);
+  
+  q.submit([&](sycl::handler &h) {
+    using group_load = dpct::group::workgroup_load<4, T, int,  int *, sycl::nd_item<3>>;
+    size_t temp_storage_size = group_load::get_local_memory_size(128);
+    sycl::local_accessor<uint8_t, 1> tacc(sycl::range<1>(temp_storage_size), h);
+    sycl::accessor data_accessor(buffer, h, sycl::read_write);
+    
+    h.parallel_for(
+        sycl::nd_range<3>(sycl::range<3>(1, 1, 128), sycl::range<3>(1, 1, 128)),
+        [=](sycl::nd_item<3> item) {
+          int thread_data[4];
+          auto *d = data_accessor.get_multi_ptr<sycl::access::decorated::yes>().get();
+          auto *tmp = tacc.get_multi_ptr<sycl::access::decorated::yes>().get();
+          group_load(tmp).load(item, d, thread_data);
+          item.barrier(sycl::access::fence_space::local_space);

```

Comment:
This barrier should not be necessary and is a signal that something is wrong.  
The load operation should be on the same work-id as the subsequent assignment operations, therefore no barrier should be necessary.  You should only need a barrier is if multiple threads are coordinating to fill in the thread local data, but that is not the case.  

The problem that this barrier "fixes" in these tests is that you are loading from the same buffer that you are then writing to as output.  The barrier separates the load from the store.  While this passes, I don't think it is the way we want to portray that these group load functions should be used because the extra synchronization is a big performance penalty, and in normal practice shouldn't be necessary.

The subgroup striped version has neither a barrier nor an output buffer.  It does still pass, so I think it is "getting lucky" currently, but should be shifted to use an output buffer.

Please add a output data buffer / accessor etc. to all 3 of these these `parallel_for` calls, and remove the barriers.  

It may be good to change the buffers from `read_write` to `read_only` and `write_only` for input and output respectively.  However, if you do this you will need to also change the `int *` in the `dpct::group::workgroup_load` definition to be `const int *`.